### PR TITLE
Mj/roaming spec

### PR DESCRIPTION
**Pull request by michaeldjeffrey**
This PR contains the base implementation for the http roaming spec.
It continues support for the UDP protocol, while testing with http can start.

To come in upcoming PRs
- [ ] authentication
- [ ] configurable incoming endpoint
- [ ] configurable outgoing endpoint for roaming
- [ ] metrics

### CODE REVIEW

**Comment by lthiery** [comment]
:eyes:

File: src/pp_config.erl [hunk 1]
```diff
@@ -377,26 +380,34 @@ transform_config(ConfigList0) ->
...
         case maps:get(<<"multi_buy">>, Entry, null) of
             null -> unlimited;
             Val -> Val
         end,
     Joins = maps:get(<<"joins">>, Entry, []),
     DisablePullData = maps:get(<<"disable_pull_data">>, Entry, false),
+
+    Protocol0 = maps:get(<<"protocol">>, Entry, ?DEFAULT_PROTOCOL),
+    Protocol =
+        case erlang:binary_to_existing_atom(Protocol0) of
```
Comment (macpie) [inline src/pp_config.erl:394]: why not just match on the binary here?

**Comment by macpie** [comment]
Can we attempt to use new lora library instead of importing old code?

**Comment by michaeldjeffrey** [comment]
> Can we attempt to use new lora library instead of importing old code?

That's on my radar. It's not quite ready to be pulled in yet.

File: src/pp_roaming_protocol.erl [hunk 2]
```diff
@@ -0,0 +1,255 @@
+-module(pp_roaming_protocol).
+
+%% Uplinking
+-export ([make_uplink_payload/3]).
+
+%% Downlinking
+-export([handle_prstart_ans/1,
```
Comment (macpie) [inline src/pp_roaming_protocol.erl:7]: ewww

**Review by macpie — APPROVED** [review]

File: src/pp_roaming_protocol.erl (hunk 2 above)
Comment (michaeldjeffrey) [reply src/pp_roaming_protocol.erl]: agree 🤮

**Pull request merged**